### PR TITLE
Improvements related to lexer

### PR DESCRIPTION
**Pull request by ktorpi**
Improvements related to lexer: 
* duplicated clone elimination
* magic numbers are replaced with constants
* functions are moved form util to lit-char-helpers

JerryScript-DCO-1.0-Signed-off-by: István Kádár ikadar@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-lexer.c [hunk 1]
```diff
@@ -1775,7 +1763,7 @@ lexer_construct_regexp_object (parser_context_t *context_p, /**< context */
   }
 
   if (source_p < source_end_p
-      && util_is_identifier_part (source_p))
+        && lit_char_is_identifier_part (source_p))
```
Comment (LaszloLango) [inline jerry-core/parser/js/js-lexer.c]: wrong indentation

File: jerry-core/lit/lit-char-helpers.c [hunk 2]
```diff
@@ -283,6 +339,61 @@ lit_char_hex_to_int (ecma_char_t c) /**< code unit, corresponding to
...
+  if (!(chr & ~LIT_UTF8_2_BYTE_CODE_POINT_MAX))
+  {
+    /* 00000yyy yyxxxxxx -> 110yyyyy 10xxxxxx */
+    *(dst_p++) = (uint8_t) (LIT_UTF8_EXTRA_BYTE_MASK | ((chr >> 6) & LIT_UTF8_LAST_5_BITS_MASK));
+    *dst_p = (uint8_t) (LIT_UTF8_1_BYTE_MASK | (chr & LIT_UTF8_LAST_6_BITS_MASK));
+    return 2;
+  }
+
+  /* zzzzyyyy yyxxxxxx -> 1110zzzz 10yyyyyy 10xxxxxx */
+  *(dst_p++) = (uint8_t) (LIT_UTF8_2_BYTE_MASK | ((chr >> 12) & LIT_UTF8_LAST_4_BITS_MASK));
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: `(LIT_UTF8_3_BYTE_MARKER | ((chr >> 12) & LIT_UTF8_LAST_4_BITS_MASK))`

File: jerry-core/lit/lit-char-helpers.c [hunk 3]
```diff
@@ -283,6 +339,61 @@ lit_char_hex_to_int (ecma_char_t c) /**< code unit, corresponding to
...
+  {
+    /* 00000000 0xxxxxxx -> 0xxxxxxx */
+    *dst_p = (uint8_t) chr;
+    return 1;
+  }
+
+  if (!(chr & ~LIT_UTF8_2_BYTE_CODE_POINT_MAX))
+  {
+    /* 00000yyy yyxxxxxx -> 110yyyyy 10xxxxxx */
+    *(dst_p++) = (uint8_t) (LIT_UTF8_EXTRA_BYTE_MASK | ((chr >> 6) & LIT_UTF8_LAST_5_BITS_MASK));
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: (LIT_UTF8_2_BYTE_MARKER | ((chr >> 6) & LIT_UTF8_LAST_5_BITS_MASK));

File: jerry-core/lit/lit-char-helpers.c [hunk 4]
```diff
@@ -283,6 +339,61 @@ lit_char_hex_to_int (ecma_char_t c) /**< code unit, corresponding to
...
+    *dst_p = (uint8_t) (LIT_UTF8_1_BYTE_MASK | (chr & LIT_UTF8_LAST_6_BITS_MASK));
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: `(LIT_UTF8_EXTRA_BYTE_MARKER | (chr & LIT_UTF8_LAST_6_BITS_MASK))`

File: jerry-core/lit/lit-char-helpers.c [hunk 5]
```diff
@@ -283,6 +339,61 @@ lit_char_hex_to_int (ecma_char_t c) /**< code unit, corresponding to
...
+    return 2;
+  }
+
+  /* zzzzyyyy yyxxxxxx -> 1110zzzz 10yyyyyy 10xxxxxx */
+  *(dst_p++) = (uint8_t) (LIT_UTF8_2_BYTE_MASK | ((chr >> 12) & LIT_UTF8_LAST_4_BITS_MASK));
+  *(dst_p++) = (uint8_t) (LIT_UTF8_1_BYTE_MASK | ((chr >> 6) & LIT_UTF8_LAST_6_BITS_MASK));
+  *dst_p = (uint8_t) (LIT_UTF8_1_BYTE_MASK | (chr & LIT_UTF8_LAST_6_BITS_MASK));
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: likewise

File: jerry-core/lit/lit-char-helpers.h [hunk 6]
```diff
@@ -210,10 +212,14 @@ extern bool lit_char_is_unicode_non_letter_ident_part (ecma_char_t);
 #define LIT_CHAR_ASCII_DIGITS_BEGIN                 LIT_CHAR_0           /* decimal digits range */
 #define LIT_CHAR_ASCII_DIGITS_END                   LIT_CHAR_9
 
+#define LEXER_TO_ASCII_LOWERCASE(character) ((character) | 0x20)
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.h]: use LEXER_FLAG_SOURCE_PTR

Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.h]: Maybe `LIT_CHAR_SP` is suitable

File: jerry-core/lit/lit-char-helpers.c [hunk 7]
```diff
@@ -283,6 +339,61 @@ lit_char_hex_to_int (ecma_char_t c) /**< code unit, corresponding to
...
+  }
+
+  if (!(chr & ~LIT_UTF8_2_BYTE_CODE_POINT_MAX))
+  {
+    /* 00000yyy yyxxxxxx */
+    return 2;
+  }
+
+  /* zzzzyyyy yyxxxxxx */
+  return 3;
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.c]: Please add an assert here (it is really 3 byte long)

File: jerry-core/lit/lit-char-helpers.c [hunk 8]
```diff
@@ -283,6 +339,61 @@ lit_char_hex_to_int (ecma_char_t c) /**< code unit, corresponding to
...
+    *(dst_p++) = (uint8_t) (LIT_UTF8_2_BYTE_MARKER | ((chr >> 6) & LIT_UTF8_LAST_5_BITS_MASK));
+    *dst_p = (uint8_t) (LIT_UTF8_EXTRA_BYTE_MARKER | (chr & LIT_UTF8_LAST_6_BITS_MASK));
+    return 2;
+  }
+
+  /* zzzzyyyy yyxxxxxx -> 1110zzzz 10yyyyyy 10xxxxxx */
+  *(dst_p++) = (uint8_t) (LIT_UTF8_3_BYTE_MARKER | ((chr >> 12) & LIT_UTF8_LAST_4_BITS_MASK));
+  *(dst_p++) = (uint8_t) (LIT_UTF8_EXTRA_BYTE_MARKER | ((chr >> 6) & LIT_UTF8_LAST_6_BITS_MASK));
+  *dst_p = (uint8_t) (LIT_UTF8_EXTRA_BYTE_MARKER | (chr & LIT_UTF8_LAST_6_BITS_MASK));
+  return 3;
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.c]: Similar assert here.

**Comment by zherczeg** [comment]
Good patch.

**Comment by LaszloLango** [comment]
Please rebase to the current master

File: jerry-core/parser/js/js-lexer.c [hunk 9]
```diff
@@ -69,7 +69,7 @@ lexer_hex_to_character (parser_context_t *context_p, /**< context */
     }
     else
     {
-      byte |= 0x20;
+      byte = LEXER_TO_ASCII_LOWERCASE (byte);
       if (byte >= 'a' && byte <= 'f')
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c]: There is a macro for 'a' and 'f' as far as I remember.

File: jerry-core/parser/js/js-lexer.c [hunk 10]
```diff
@@ -779,13 +768,13 @@ lexer_parse_number (parser_context_t *context_p) /**< context */
   if (source_p[0] == '0'
       && source_p + 1 < source_end_p)
   {
-    if ((source_p[1] | 0x20) == 'x')
+    if (LEXER_TO_ASCII_LOWERCASE (source_p[1]) == 'x')
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c]: Change 'x' to constant.

**Comment by zherczeg** [comment]
Please change all 'letter' constants to symbolic constants. After that, LGTM.

File: jerry-core/parser/js/js-lexer.c [hunk 11]
```diff
@@ -1112,27 +1101,37 @@ lexer_next_token (parser_context_t *context_p) /**< context */
...
+    LEXER_TYPE_C_TOKEN (LIT_CHAR_VLINE, LEXER_BIT_OR, LIT_CHAR_EQUALS,
+                        LEXER_ASSIGN_BIT_OR, LIT_CHAR_VLINE, LEXER_LOGICAL_OR)
+
+    LEXER_TYPE_B_TOKEN (LIT_CHAR_CIRCUMFLEX, LEXER_BIT_XOR, LIT_CHAR_EQUALS,
+                        LEXER_ASSIGN_BIT_XOR)
+
+    LEXER_TYPE_A_TOKEN (LIT_CHAR_TILDE, LEXER_BIT_NOT);
+    LEXER_TYPE_A_TOKEN (LIT_CHAR_QUESTION, LEXER_QUESTION_MARK);
+    LEXER_TYPE_A_TOKEN (LIT_CHAR_COLON, LEXER_COLON);
+
```
Comment (LaszloLango) [inline jerry-core/parser/js/js-lexer.c]: Missing comment: `/* FALLTHRU */`

Comment (LaszloLango) [inline jerry-core/parser/js/js-lexer.c]: My bad. BTW, it is hard to read this part. The macro name should refer to it makes a `switch-case`

File: jerry-core/parser/js/js-lexer.c [hunk 12]
```diff
@@ -1476,17 +1475,17 @@ lexer_construct_literal_object (parser_context_t *context_p, /**< context */
...
 
     if (literal_p->length == 4
-        && source_p[0] == 'e'
-        && source_p[3] == 'l'
-        && source_p[1] == 'v'
-        && source_p[2] == 'a')
+        && source_p[0] == LIT_CHAR_LOWERCASE_E
+        && source_p[3] == LIT_CHAR_LOWERCASE_L
+        && source_p[1] == LIT_CHAR_LOWERCASE_V
+        && source_p[2] == LIT_CHAR_LOWERCASE_A)
```
Comment (LaszloLango) [inline jerry-core/parser/js/js-lexer.c]: Strange order. Has this any benefit?

File: jerry-core/parser/js/js-lexer.c [hunk 13]
```diff
@@ -1476,17 +1475,17 @@ lexer_construct_literal_object (parser_context_t *context_p, /**< context */
...
     {
       context_p->lit_object.type = LEXER_LITERAL_OBJECT_EVAL;
     }
 
     if (literal_p->length == 9
-        && source_p[0] == 'a'
-        && source_p[8] == 's'
+        && source_p[0] == LIT_CHAR_LOWERCASE_A
+        && source_p[8] == LIT_CHAR_LOWERCASE_S
         && memcmp (source_p + 1, "rgument", 7) == 0)
```
Comment (LaszloLango) [inline jerry-core/parser/js/js-lexer.c]: Is this really faster then `memcmp (source_p + 1, "arguments", 9) == 0`?

File: jerry-core/parser/js/js-lexer.c (hunk 12 above)
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c]: Yes, when the first and last character matches, there is a high chance there will be a match.

**Comment by LaszloLango** [comment]
LGTM

**Comment by LaszloLango** [comment]
@zherczeg, may we land this?

**Comment by zherczeg** [comment]
Yes of course, LGTM.

**Pull request merged**